x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -536,8 +539,14 @@ private static String getAction(BulkItemRequest item) {
         final DocWriteRequest<?> docWriteRequest = item.request();
         switch (docWriteRequest.opType()) {
             case INDEX:
+                IndexRequest request = (IndexRequest) item.request();
+                if (request.getAutoGeneratedTimestamp() != IndexRequest.UNSET_AUTO_GENERATED_TIMESTAMP) {
```

Comment:
I wonder if we should make the `opType()` in IndexRequest return CREATE in case of auto-generated IDs. This  would simplify the logic here. Can you prototype that change in the `IndexRequest` class? You can draw inspiration from the `resolveVersionDefaults` method and override the `opType` getter in a similar way:

```
    public OpType opType() {
        if (autoGeneratedTimestamp != UNSET_AUTO_GENERATED_TIMESTAMP) {
            return OpType.CREATE;
        } else {
            return this.opType;
        }
    }
```

Reply:
I think that this change in behavior is ok, especially given that this is the behavior this privilege is being built upon. Wouldn't it be odd for the "create_docs" privilege to allow overriding docs just because it was an append-only index request?

> Agreed there is a less chance of an auto-generated ID conflict but if a document with that ID exists then by old behavior it would have just updated the document, but if we change the behavior to return CREATE then it will start throwing exceptions. Is that okay?

I think that this is ok.

The only conflict that can happen is due to internal retries of an operation if there is a disconnect between nodes in a cluster or if there is a primary failover. In these rare cases, it could now lead to a case where an append-only operation results in a `VersionConflictException`. I'm not too concerned about this case.

The bigger concern I have with my proposal is that it changes the API response. An append-only indexing operation will now return "create" in its API response instead of "index". You could argue that this is a breaking change. We could of course work around this by returning "index" for now, and only adapt the response in 8.0.

Would be interesting to get other folks' thoughts here. Marking it as team-discuss if you don't mind.

Reply:
> > Agreed there is a less chance of an auto-generated ID conflict but if a document with that ID exists then by old behavior it would have just updated the document, but if we change the behavior to return CREATE then it will start throwing exceptions. Is that okay?
>>
> I think that this is ok.

I also think this is a positive side-effect of this change

Reply:
Elasticsearch 7.5 now supports bulk requests with auto-generated ID using op_type create, and single index requests (`POST indexName/_doc`) now also default to op_type create. This means that from an ES point of view we can now directly associate the "create_doc" index privilege with the op_type create and can get rid of the exception in the privilege code here. The only thing worth noting is that Beats defaults to op_type index for bulk requests with auto-generated id (as this was not supported in ES prior to 7.5). We need an adaptation in Beats to make it work with the amended index privilege. Logstash also currently requires (independently whether auto-generated IDs are used or not) the user to explicitly configure op_type create to work with the amended create_doc index privilege.